Partial fix for #2381: clean up sources of memory leaks due to '$on' and '$watch'.

seanlip opened the pull request:
This follows guidance in http://www.dwmkerr.com/fixing-memory-leaks-in-angularjs-applications/#fixingmemoryleaks (see issue #2381 for more details).

seanlip commented:
@BenHenning -- mind taking a look (if time permits)? Thanks!

seanlip commented:
@BenHenning ping (if you have a moment!)

BenHenning commented on core/templates/dev/head/components/AnswerGroupEditorDirective.js:
```diff
@@ -239,13 +225,31 @@ oppia.directive('answerGroupEditor', [function() {
           $scope.answerChoices = $scope.getAnswerChoices();
         });
+
+        $scope.$on('$destroy', cleanup1);
```
I'm not keen on variables named cleanup, or the requirement to do this extra step everytime we want to register a listener for a globally broadcasted Angular message. Can we add some sort of utility function shared among the entire Oppia module to pipeline this for us? Maybe something like:

```javascript
BroadcastService.on($scope, 'updateAnswerChoices', function() {
  ...
});
```

Where ``BroadcastService.on`` would look something like:

```javascript
on: function($scope, eventName, callback) {
  var cleanup = $scope.$on(eventName, callback);
  $scope.$on('$destroy', cleanup);
}
```

It would be nicer if we could add a custom function to ``$scope`` itself, but I don't know how feasible that is.

BenHenning commented on core/templates/dev/head/components/OutcomeDestinationEditorDirective.js:
```diff
@@ -44,20 +57,10 @@ oppia.directive('outcomeDestinationEditor', [function() {
-        $scope.newStateNamePattern = /^[a-zA-Z0-9.\s-]+$/;
-        $scope.destChoices = [];
-        $scope.$watch(explorationStatesService.getStates, function() {
+        var cleanup2 = $scope.$watch(
```
Ditto for this. Can we have a ``BroadcastService.watch``, too?

BenHenning commented on core/templates/dev/head/directives.js:
```diff
@@ -96,6 +97,11 @@ oppia.directive('customPopover', ['$sce', function($sce) {
         $scope.isShown = false;
       });
 
+      $scope.$on('$destroy', function() {
```
Can this just be: ``$scope.$on('$destroy', $element.off);`` ?

seanlip replied:
Done.

BenHenning commented on core/templates/dev/head/directives.js:
```diff
@@ -159,6 +165,11 @@ oppia.directive('mobileFriendlyTooltip', ['$timeout', function($timeout) {
         });
       };
+
+      scope.$on('$destroy', function() {
```
Ditto, but for ``element`` (and elsewhere that uses this pattern)

seanlip replied:
Done.

BenHenning commented on core/templates/dev/head/domain/dashboard/ScrollOnClickDirective.js:
```diff
@@ -1,3 +1,22 @@
+// Copyright 2015 The Oppia Authors. All Rights Reserved.
```
Nit: 2016

BenHenning commented on extensions/objects/templates/UnicodeStringEditor.js:
```diff
@@ -33,25 +33,28 @@ oppia.directive('unicodeStringEditor', [
+        var cleanup3 = $scope.$watch('localValue.label', function(newValue) {
           $scope.$parent.value = newValue;
         });
+        $scope.$on('$destroy', cleanup1);
```
Should be ``cleanup3`` I believe. I think this reinforces my suggestion about how this approach is error prone. :)

seanlip replied:
Good catch! Fixed temporarily, pending reply to general refactor.

seanlip commented:
Hi @BenHenning, thanks!

I suspect that there is more performance work to be done, but brief tests I did with taking heap snapshots suggest that this is indeed clearing up some memory. So it seems like a "might as well do it" thing, which is consistent with some of the advice I've been finding on the Web.

I've addressed all your comments except for the two that involve creating a new service. Please could you take a look at the following files:

- core/templates/dev/head/components/AnswerGroupEditorDirective.js  
- core/templates/dev/head/components/OutcomeDestinationEditorDirective.js 
- core/templates/dev/head/services/CleanupService.js 

and let me know if you are happy with the approach I took? If so, I will replicate these changes for the other files and then this PR should be ready for submission.

Thanks!

BenHenning commented:
Discussed offline with @seanlip. Closing this since it doesn't seem to be a viable solution to the problem. According to Angular source code, the [deregister function returned by ``$on``](https://github.com/angular/angular.js/blob/406c1b094bc0a8a79b06f0d5f0cecdfb9f6087b0/src/ng/rootScope.js#L1162) does an internal cleanup mechanism for the listener. This same mechanism happens as part of the scope's [``$destroy``](https://github.com/angular/angular.js/blob/406c1b094bc0a8a79b06f0d5f0cecdfb9f6087b0/src/ng/rootScope.js#L923) function. And, that function is also responsible for firing the ``$destroy`` event, which in turn calls the cleanups per the proposed ``CleanupService``. This cleanup service won't actually solve any problems Angular doesn't handle by itself.

Closing this issue, as agreed to by @seanlip. Thanks!

seanlip commented:
For posterity, also noting that we decided to change the name 'CleanupService' to 'AngularService' before deciding to dispense with this PR.
